Use UTC in Timestamps when authenticating against Emarsys

or opened the pull request:
This was the only extra commit in https://github.com/eisensheng/python-emarsys/
used by https://github.com/machtfit/signup.

https://github.com/machtfit/signup is going to reference the machtfit repository in the future.

Azd325 commented on emarsys/api.py:
```diff
@@ -111,8 +111,10 @@ def call(self, uri, method, params=None):  # noqa
     def _authentication_header_value(self):
-        now = datetime.datetime.now(self._tzinfo_obj)
-        created = now.replace(microsecond=0).isoformat()
+        created = (datetime.datetime.utcnow()
```
Vielleicht so`datetime.datetime.now(pytz.utc).isoformat()`
https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow

or replied:
👍

Azd325 commented on emarsys/api.py:
```diff
@@ -111,8 +111,10 @@ def call(self, uri, method, params=None):  # noqa
-        now = datetime.datetime.now(self._tzinfo_obj)
-        created = now.replace(microsecond=0).isoformat()
+        created = (datetime.datetime.utcnow()
+                   .replace(microsecond=0, tzinfo=pytz.utc)
```
Warum werden die microsecond weg gemacht?

or replied:
Vermutlich weil die API solche timestamps erwartet. Da das so aber jahrelang funktioniert hat, möchte ich daran im Moment nichts ändern.

Azd325 commented on Pipfile:
```diff
@@ -13,6 +13,7 @@ responses = "*"
 "flake8" = "*"
 isort = "*"
 tox = "*"
+pytz = "*"
```
`pipenv lock` müsstes du einmal aufrufen das die Lockdatei noch erzeugt wird.

or replied:
Done.

Azd325 commented on tests/test_emarsys.py:
```diff
@@ -12,9 +12,9 @@ def test_create_authentication_header(monkeypatch):
     assert value == ('UsernameToken'
                      ' Username="client_id",'
-                     ' PasswordDigest="YzI3MWVhMDgwNjMzZDdhMjE1MjhkMzM3NTk1NTQwYjRjNmM4ZGYzMA==",'
+                     ' PasswordDigest="NzVhOTg3Yjg2NTcyZmE1OWMwYzIxNWQ5ZDgyMzgxZWExOGI3NGFjZQ==",'
```
Vorher kommt dieser neue Wert?
Wenn es einen Grund hat würde ich diesen vllt separieren.

or replied:
War im ursprünglichen commit drin. Er muß sich sicher wegen des neuen timestamps ändern. Gehört aber definitiv zum commit.

Azd325 commented on tox.ini:
```diff
@@ -1,7 +1,6 @@
 [tox]
 envlist = py27, py36, py37 flake8-py3
 
-
```
Müsste nicht unbedingt hier drin sein. :)

or replied:
Rausgenommen.

Azd325 reviewed (commented):
Warum brauchen wir utc für emarsys authentication? Das wird nirgend wirklich aufgeklärt. :)

or commented:
> Warum brauchen wir utc für emarsys authentication? Das wird nirgend wirklich aufgeklärt. :)

Das weiß ich leider auch nicht. Es funktioniert offensichtlich in machtfit-oscar auch ohne. Aber signup lief nun seit Jahren mit diesem timestamp, deshalb möchte ich den derzeitigen Zustand beibehalten.

Azd325 commented:
https://dev.emarsys.com/v2/before-you-start/authentication#timezone-considerations

or commented:
> https://dev.emarsys.com/v2/before-you-start/authentication#timezone-considerations

Cool, sollte also auch für machtfit-oscar so funktionieren. Explizit die Zeitzone angeben ist sicher so oder so besser.

Azd325 commented:
Vielleicht können wir den Link zum Commit hinzufügen.

Azd325 reviewed (approved):
LGTM

mbertheau commented on emarsys/api.py:
```diff
@@ -111,8 +111,10 @@ def call(self, uri, method, params=None):  # noqa
-        created = now.replace(microsecond=0).isoformat()
+        created = (datetime.datetime.now(pytz.utc)
+                   .replace(microsecond=0)
+                   .isoformat())
```
Ist jetzt hier der Unterschied, das das dem Konstruktor übergebene `tzinfo_obj` ignoriert und UTC hartkodiert wird? Wenn ja, dann sollte das so in der Commit-Message stehen.

Ggfs. wäre es angebracht dann auch `self._tzinfo_obj` zu entfernen.

Benutzen wir bei machtfit-oscar den gleichen falschen nicht-UTC-Code?

or replied:
Ja, der originale commit hat ` tzinfo_obj` ignoriert. Ich habe nun meinen Plan verworfen, den Ist-Stand möglichst zu übernehmen, denn hier können wir wahrscheinlich die Auswirkungen noch gut kontrollieren.

Deshalb nun ein ganz neuer commit, der den originalen commit referenziert, den Code leicht ändert, und `tzinfo_obj` entfernt. Da wir `tzinfo_obj` in machtfit/signup und machtfit/machtfit-oscar nie benutzt haben, ist diese Änderung aus Sicht der beiden repositories äquivalent zum anderen commit.

Damit schon die zweite Frage beantwortet: machtfit-oscar nutzt auch die Version ohne UTC, ist jedoch derzeit noch auf python-emarsys==0.6 gepinnt. Ich glaube, der UTC default ist auch für machtfit-oscar sinnvoll, aber das können wir später anpassen.

or committed:
Bump version to 0.7

or committed:
Use UTC timestamps when authenticating with Emarsys

If Emarsys() was used without specifying a timezone, then the
resulting datetime object was not timezone aware. A timestamp without
timezone information is interpreted as a timestamp in the local
timezone of the Emarsys server (GMT+1 or GMT+2 during DST).
https://dev.emarsys.com/v2/before-you-start/authentication#timezone-considerations

I don't know the history of the original commit below, but it's
conceivable that relying on the default caused issues in some edge
cases. Either way it is a good idea to be explicit about the timezone,
and since we don't specify a timezone in our code and have no good
reason to do so, we can always use UTC and remove the tzinfo_obj.

This differs from the original commit, but results in an equivalent
change from the point of view of machtfit/signup (via django-emarsys).

Originally implemented thus:
https://github.com/eisensheng/python-emarsys/commit/aaa1085f879b29f15ff860766b0ae839914453e7

https://machtfit.atlassian.net/browse/MAC-2448
